### PR TITLE
Rebuild the wrapper widget architecture

### PR DESCRIPTION
**Pull request by moritzmorgenroth**
@aissat @Overman775 I found the time today and rebuild the architecture layer of the top level wrapper widget. I eliminated the `Intl` stuff completly from the config, since I believe it is much safer / more maintainable to just trust the flutter core framework to resolve the device language config and work through the delegate, offering an override logic through the safe persistance. All features are working as expected & tests are passing. Here are the notables: 
* The initialisation of the wrapper is now happening asynchronously, meaning that no more loader or err or widgets are needed, but it will simply refreshes if a persistet locale is found. Otherwise, it is left to the system to select the locale as usual and the library will adapt to it. This will eliminate #39 #98 and #96. Also better architecture not to block anything with async work before the actual app widget. 
* I eliminated the fallback locale, since there is more advanced login to selecting it. I think this should be left up to the system. Generally, according to flutter docs, the first entry of the array will be used as fallback. However, in some cases some translation delegates might not be supporting a fallback or the system has other settings that determine preferred language. I think it is unsafe to force the fallback onto the system through a direct override. 
* I commented out the Intl. Related tests. Im not sure, what the `Intl.defaultLocale` setting actually does. Everything works fine without it. Am I overlooking something here? If so, ping me and ill set it in my implementation. 
* I would consider defaulting the language saving to false. 
* Doc updates are pending, wanted to get your feedback first.

Cheers, Moritz

### CODE REVIEW

**Comment by Overman775** [comment]
@moritzmorgenroth Nice 👍

File: lib/src/easy_localization_app.dart [hunk 1]
```diff
@@ -1,263 +1,149 @@
...
 
-  bool _checkInitLocale(Locale locale, Locale osLocale) {
-    // If suported locale not contain countryCode then check only languageCode
-    if (locale.countryCode == null) {
-      return (locale.languageCode == _osLocale.languageCode);      
-    } else {
-      return (locale == _osLocale);
+  loadSavedLocale() async {
+    SharedPreferences _preferences = await SharedPreferences.getInstance();
+    var _codeLang = _preferences.getString('codeLa');
```
Comment (Overman775) [inline lib/src/easy_localization_app.dart]: I find bug, in new version @aissat use different sharedPreference val(locale), look _saveLocale function --> sharedPreference

Comment (moritzmorgenroth) [reply lib/src/easy_localization_app.dart]: Good catch, I broke the test for taht. Sorry

**Comment by aissat** [comment]
@moritzmorgenroth great work ;), i like it

File: lib/src/easy_localization_app.dart (hunk 1 above)
Comment (moritzmorgenroth) [reply lib/src/easy_localization_app.dart]: I will fix this later tonight, with the share preference rename the language loading broken. Ill push something later.

**Comment by aissat** [comment]
@moritzmorgenroth 
```
[MyApp# locale] null
[log] easy localization: Locale loaded from shared preferences ru_RU
[MyApp# locale] ar_DZ
```

File: lib/src/easy_localization_app.dart [hunk 2]
```diff
@@ -2,118 +2,103 @@ import 'dart:developer';
...
 
-  // Get Device Locale
-  Future<Locale> _getDeviceLocale() async {
-    final String _deviceLocale = await findSystemLocale();
-    return _localeFromString(_deviceLocale);
+  @override
+  Widget build(BuildContext context) {
+    return EasyLocalizationProvider(
+      widget,
+      this.locale,
```
Comment (aissat) [inline lib/src/easy_localization_app.dart]: `this.locale`  return device locale when reload app
android output: 
```
[MyApp# locale] null
[log] easy localization: Locale loaded from shared preferences ru_RU
[MyApp# locale] ru_RU
[MyApp# locale] en_US
```

File: test/easy_localization_widget_test.dart [hunk 3]
```diff
@@ -542,80 +540,13 @@ group('SharedPreferences saveLocale', () {
...
         });
       },
     );
 
   });
-
-
-
-
-group('SharedPreferences deleteSaveLocale()', () {
```
Comment (aissat) [inline test/easy_localization_widget_test.dart]: why removed this test?

File: lib/src/easy_localization_app.dart (hunk 1 above)
Comment (aissat) [reply lib/src/easy_localization_app.dart]: don't worry about it

File: lib/src/easy_localization_app.dart (hunk 2 above)
Comment (aissat) [reply lib/src/easy_localization_app.dart]: @Overman775 plz check this

File: lib/src/easy_localization_app.dart [hunk 4]
```diff
@@ -1,263 +1,149 @@
...
-      locale = _savedLocale;      
-    } else {
-      locale = supportedLocales.firstWhere((locale) => _checkInitLocale(locale, _osLocale),
-          orElse: () => _getFallbackLocale(supportedLocales, fallbackLocale));
-    }
-    //Set locale
-    if (Intl.defaultLocale == null) locale = _locale;    
+  @override
+  void initState() {
+    loadSavedLocale();
```
Comment (Overman775) [inline lib/src/easy_localization_app.dart]: if (widget.saveLocale) loadSavedLocale();

File: test/easy_localization_widget_test.dart (hunk 3 above)
Comment (moritzmorgenroth) [reply test/easy_localization_widget_test.dart]: that was a merge issue

Comment (moritzmorgenroth) [reply test/easy_localization_widget_test.dart]: putting them back!

File: lib/src/easy_localization_app.dart (hunk 2 above)
Comment (moritzmorgenroth) [reply lib/src/easy_localization_app.dart]: @aissat this is fixed in my last commit.

File: lib/src/easy_localization_app.dart (hunk 1 above)
Comment (moritzmorgenroth) [reply lib/src/easy_localization_app.dart]: Fixed!

**Comment by moritzmorgenroth** [comment]
@aissat @Overman775 the shared preference stuff is fixed & working fine in the app, however, the tests are still failing. Do you have any idea why?

**Comment by Overman775** [comment]
00:02 +42 -1: Some tests failed.

**Comment by Overman775** [comment]
@moritzmorgenroth @aissat next bug, widget rebuild twice, first witch null locale, next normal. 
I have not found a solution yet

```
[MyApp# locale] null
[MyApp# locale] en_US
```

**Comment by moritzmorgenroth** [comment]
> @moritzmorgenroth @aissat next bug, widget rebuild twice, first witch null locale, next normal.
> I have not found a solution yet
> 
> ```
> [MyApp# locale] null
> [MyApp# locale] en_US
> ```

@Overman775 that is intendet behavior. The widget will rebuild once the delegate is told to load the first time with the correct locale. This way, we can trust the material app to select the correct language from the system. You can understand `locale == null` as a loading state.

**Comment by Overman775** [comment]
> @Overman775 no idea why, these are the tests with the shared preferences. Seems like they are not getting picket up. Works fine in the actual app. Do you have any thoughts on it?

@moritzmorgenroth I don’t know... @aissat knows better testing, maybe he will tell you.

**Comment by Overman775** [comment]
@moritzmorgenroth I don’t see in the code where the device locale is taken, tell me how?

**Comment by moritzmorgenroth** [comment]
> @moritzmorgenroth I don’t see in the code where the device locale is taken, tell me how? Flutter magic? :))

Hahahaha yea nothing like flutter magic XD 

By passing the null locale to the MaterialWidget, it resolves the system locale by itself and calls the delegate, from there its propagated back to the main widget via a callback

**Comment by aissat** [comment]
@moritzmorgenroth  @Overman775 I don't know why saveLocale not working when reloading the app
first, run:
```
Restarted application in 1,605ms.
[MyApp# locale] null
[log] easy localization: Locale loaded from shared preferences en_US
[MyApp# locale] ar_DZ
[LanguageView] ru_RU
[log] easy localization: Locale set ru_RU
[log] easy localization: Locale saved ru_RU
[MyApp# locale] ru_RU
```
reloaded

```
Restarted application in 1,909ms.
[MyApp# locale] null
[log] easy localization: Locale loaded from shared preferences ru_RU
[MyApp# locale] ar_DZ
```

**Comment by aissat** [comment]
@moritzmorgenroth @Overman775  this is crazy

**Comment by aissat** [comment]
> > 00:02 +42 -1: Some tests failed.
> 
> @Overman775 no idea why, these are the tests with the shared preferences. Seems like they are not getting picket up. Works fine in the actual app. Do you have any thoughts on it?

about testes i'll fixed

**Comment by moritzmorgenroth** [comment]
> @moritzmorgenroth @Overman775 I don't know why saveLocale not working when reloading the app
> first, run:
> 
> ```
> Restarted application in 1,605ms.
> [MyApp# locale] null
> [log] easy localization: Locale loaded from shared preferences en_US
> [MyApp# locale] ar_DZ
> [LanguageView] ru_RU
> [log] easy localization: Locale set ru_RU
> [log] easy localization: Locale saved ru_RU
> [MyApp# locale] ru_RU
> ```
> 
> reloaded
> 
> ```
> Restarted application in 1,909ms.
> [MyApp# locale] null
> [log] easy localization: Locale loaded from shared preferences ru_RU
> [MyApp# locale] ar_DZ
> ```

Ill look into it

**Comment by Overman775** [comment]
> @moritzmorgenroth @Overman775 this is crazy

```
E/AccessibilityBridge(12939): VirtualView node must not be the root node.
Restarted application in 776ms.
[MyApp# locale] null
[log] easy localization: Locale loaded from shared preferences ar_DZ
[MyApp# locale] ar_DZ
```

hotreload

```
[MyApp# locale] null
[log] easy localization: Locale loaded from shared preferences ar_DZ
[MyApp# locale] en_US
```

**Comment by aissat** [comment]
> > @moritzmorgenroth @Overman775 this is crazy
> 
> ```
> E/AccessibilityBridge(12939): VirtualView node must not be the root node.
> Restarted application in 776ms.
> [MyApp# locale] null
> [log] easy localization: Locale loaded from shared preferences ar_DZ
> [MyApp# locale] ar_DZ
> ```
> 
> hotreload
> 
> ```
> [MyApp# locale] null
> [log] easy localization: Locale loaded from shared preferences ar_DZ
> [MyApp# locale] en_US
> ```

@Overman775 @moritzmorgenroth  thinking this pbm form `async` func

**Comment by aissat** [comment]
@moritzmorgenroth @Overman775 I think the solutions are available is:
- init `loadSavedLocale` in `main()` func like @spiritinlife  approach 
```dart
void main() async {
  await EasyLocalization.ensureInitialized();
```
- used  `then`some thing lik this 
``` dart
loadAsyncData().then((result) {
            // If we need to rebuild the widget with the resulting data,
            // make sure to use `setState`
            setState(() {
                _result = result;
            });
        }); 
```
- used `FutureBuilder`
- used `StreamBuilder` BLOC

**Comment by moritzmorgenroth** [comment]
@aissat let me look into it tomorrow. It has something todo with the loading delay in the shared prefs value. I think there might be an easier solution

**Comment by aissat** [comment]
> @aissat let me look into it tomorrow. It has something todo with the loading delay in the shared prefs value. I think there might be an easier solution

@moritzmorgenroth thnx man, take ur time, I'm now working on `StreamBuilder`  tonight I'll push code

**Comment by moritzmorgenroth** [comment]
@aissat no problem, I think I found the issue. no need for a streambuilder.

**Comment by aissat** [comment]
> @aissat no problem, I think I found the issue. no need for a streambuilder.
@moritzmorgenroth  nice man

**Comment by moritzmorgenroth** [comment]
Fixed. I hope tests will pass now too :) Fingers crossed :)

**Comment by aissat** [comment]
> Fixed. I hope tests will pass now too :) Fingers crossed :)

@ moritzmorgenroth  great , lol

**Comment by aissat** [comment]
> Fixed. I hope tests will pass now too :) Fingers crossed :)
@moritzmorgenroth  great work everything working well

**Pull request merged**